xonsh/jobs.py

```diff
@@ -105,21 +106,36 @@ def _set_pgrp(info):
 
     _shell_pgrp = os.getpgrp()
 
-    _block_when_giving = (signal.SIGTTOU, signal.SIGTTIN,
-                          signal.SIGTSTP, signal.SIGCHLD)
+    _block_when_giving = LazyObject(lambda: (signal.SIGTTOU, signal.SIGTTIN,
+                                             signal.SIGTSTP, signal.SIGCHLD),
+                                    globals(), '_block_when_giving')
+
+    # check for shell tty
+    @functools.lru_cache(1)
+    def _shell_tty():
+        try:
+            _shtty = sys.stderr.fileno()
+            if os.tcgetpgrp(_shtty) != os.getpgid(os.getpid()):
+                # we don't own it
+                _shtty = None
+        except OSError:
+            _shtty = None
+        return _shtty
+
 
     # _give_terminal_to is a simplified version of:
     #    give_terminal_to from bash 4.3 source, jobs.c, line 4030
     # this will give the terminal to the process group pgid
     if ON_CYGWIN:
-        import ctypes
-        _libc = ctypes.CDLL('cygwin1.dll')
+        _libc = LazyObject(lambda: ctypes.CDLL('cygwin1.dll'),
+                           globals(), '_libc')
 
         # on cygwin, signal.pthread_sigmask does not exist in Python, even
         # though pthread_sigmask is defined in the kernel.  thus, we use
         # ctypes to mimic the calls in the "normal" version below.
         def _give_terminal_to(pgid):
-            if _shell_tty is not None and os.isatty(_shell_tty):
+            shtty = _shell_tty()
```

Comment:
I will not say this phonetically.  I will not say this phonetically.  I will not say this phonetica....